### PR TITLE
Add Module.reserved_attributes/0

### PR DESCRIPTION
**Pull request by wojtekmach**
Worth mentioning that we have also a few more standard attributes:

  - `@tag` for ExUnit
  - `@shortdoc` for Mix

but since they are not part of Elixir, they are not included in this list.

cc @msaraiva

### CODE REVIEW

File: lib/elixir/lib/module.ex [hunk 1]
```diff
@@ -543,6 +543,95 @@ defmodule Module do
...
 
+  @doc """
+  Returns information about module attributes used by Elixir.
+
+  See the "Module attributes" section in the module documentation for more
+  information on each attribute.
+
+  ## Examples
+
+      iex> map = Module.builtin_attributes()
```
Comment (eksperimental) [inline lib/elixir/lib/module.ex:554]:
```suggestion
      iex> map = Module.reserved_attributes()
```

**Review by eksperimental — CHANGES_REQUESTED** [review]

**Comment by eksperimental** [comment]
> What about `fallback_to_any`, `derive`, `enforce_keys`, `vsn`? There are probably a few other reserved ones. One i can think of is `__struct__`

I came to comment on `__struct__`, but since it starts with an underscore we may not want to list it.

**Comment by wojtekmach** [comment]
> What about fallback_to_any, derive, enforce_keys, vsn?

Good catches, thanks. These are specific to protocols & structs but perhaps they should be listed there as well. I listed just the ones that I found in `Module` documentation.

**Comment by josevalim** [comment]
I think we should add @enforce_keys and @derive for sure because it applies to any module. Not sure about @fallback_to_any. Definitely a no for __struct__ as it is private.

**Comment by eksperimental** [comment]
Protocols use the `@functions` attribute

**Comment by eksperimental** [comment]
> Protocols use the `@functions` attribute

Since this is undocumented (AFAIK) we could make it private, and name it `@__functions__`

**Comment by wojtekmach** [comment]
Just had an idea how we could do the same thing for ExUnit `@tag` etc, how about the following?

```elixir
defmodule Module do
  def reserved_attributes(module) do
    if function_exported?(...), do: module.__reserved_attributes__(), else: %{}
  end

  def reserved_attributes() do
    %{moduledoc: ..., doc: ..., ...}
  end
end

defmodule ExUnit.Case do
  def __reserved_attributes__() do
    %{tag: ..., moduletag: ..., describetag: ...}
  end
end

defmodule Ecto.Schema do
  def __reserved_attributes__() do
    %{primary_key: ..., timestamps_opts: ..., ...}
  end
end
```

and then the tooling upon seeing `use mod` would call `Module.reserved_attributes(mod)`.

Not sure if coupling to `__using__` like that is a good idea though, if we don't want to define `@enforce_keys` "globally", using this mechanism I'd put it on `Protocol.__reserved_attributes__()` but how one would automatically infer it, special case `defimpl`?

**Comment by josevalim** [comment]
@eksperimental i was talking to @msaraiva from Elixirsense and we agreed that all private attributes should start with an underscore. Maybe `@__functions__` is too much but standardizing on `@_functions` would be fine? Do you want to send a PR for that?

I don’t think we want to make this customizable for now though. So the current definition plus the structs ones are fine by me.

**Comment by wojtekmach** [comment]
> I don’t think we want to make this customizable for now though

fair enough. if we'd go with that `reserved_attributes` idea, I suppose we could add a `reserved_attributes(module \\ Module)` in a backwards compatible way :)

File: lib/elixir/lib/module.ex [hunk 2]
```diff
@@ -543,6 +543,101 @@ defmodule Module do
   @callback __info__(:md5) :: binary()
   @callback __info__(:module) :: module()
 
+  @doc """
+  Returns information about module attributes used by Elixir.
+
+  See the "Module attributes" section in the module documentation for more
+  information on each attribute.
```
Comment (wojtekmach) [inline lib/elixir/lib/module.ex:550]: I suppose this sentence is no longer true for `@derive` and `@enforce_keys`, how should we reword it? Or just remove it?

Comment (josevalim) [reply lib/elixir/lib/module.ex:550]: Let’s add a quick sentence at the top on derive and enforce keys and point people to structs. :)

**Comment by eksperimental** [comment]
> @eksperimental i was talking to @msaraiva from Elixirsense and we agreed that all private attributes should start with an underscore. Maybe `@__functions__` is too much but standardizing on `@_functions` would be fine? Do you want to send a PR for that?
> 
> I don’t think we want to make this customizable for now though. So the current definition plus the structs ones are fine by me.

I'm fine with it.
Sure I can contribute a PR

File: lib/elixir/lib/module.ex [hunk 3]
```diff
@@ -543,6 +553,103 @@ defmodule Module do
...
+  ## Examples
+
+      iex> map = Module.reserved_attributes()
+      iex> Map.has_key?(map, :moduledoc)
+      true
+      iex> Map.has_key?(map, :doc)
+      true
+
+  """
+  @doc since: "1.13.0"
```
Comment (wojtekmach) [inline lib/elixir/lib/module.ex:571]: @josevalim in case you want to backport it to v1.12 :)

```suggestion
  @doc since: "1.12.0"
```

**Comment by josevalim** [comment]
:green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:

**Comment by eksperimental** [comment]
For the records, we also have `@protocol` and `@for` that only apply to protocol implementations.

**Pull request merged**